### PR TITLE
[90] json metrics for plotting

### LINKED ISSUE #90: TrainLogger.read should use the json format
We are writing comprehensive logs with #29  . They should be used when reading in the new format.

Objective: complete the read path too.

### PR DESCRIPTION
**Pull request by tjhunter**
Closes #90 . 
This plugs the current plotting functions into the json metrics.

@clessig  I will need your help to complete the current code, the code for plotting is not quite documented.

The data is exported as a dataframe, I am unsure how to go from this structured output to the current numpy arrays. In my opinion, it should simply not go down to numpy arrays for plotting. What are your thoughts there?

```
Schema({'weathergen.timestamp': Datetime(time_unit='ms', time_zone=None), 'weathergen.time': Int64, 'num_samples': Float64, 'loss_avg_0_mean': Float64, 'learning_rate': Float64, 'stream.era5.loss_mse.loss_avg': Float64, 'stream.era5.stddev_avg': Float64, 'perf.gpu': Float64, 'perf.memory': Float64})


┌─────────────────────────┬─────────────┬─────────────────┬───────────────┬───────────────────────────────┬────────────────────────┐
│ weathergen.timestamp    ┆ num_samples ┆ loss_avg_0_mean ┆ learning_rate ┆ stream.era5.loss_mse.loss_avg ┆ stream.era5.stddev_avg │
│ ---                     ┆ ---         ┆ ---             ┆ ---           ┆ ---                           ┆ ---                    │
│ datetime[ms]            ┆ f64         ┆ f64             ┆ f64           ┆ f64                           ┆ f64                    │
╞═════════════════════════╪═════════════╪═════════════════╪═══════════════╪═══════════════════════════════╪════════════════════════╡
│ 2025-03-18 16:42:37.571 ┆ 20.0        ┆ 1.094918        ┆ 0.000001      ┆ 1.094918                      ┆ NaN                    │
│ 2025-03-18 16:43:35.623 ┆ 40.0        ┆ 1.056762        ┆ 0.000003      ┆ 1.056762                      ┆ NaN                    │
│ 2025-03-18 16:44:47.891 ┆ 60.0        ┆ 1.019803        ┆ 0.000005      ┆ 1.019803                      ┆ NaN                    │
│ 2025-03-18 16:46:03.834 ┆ 80.0        ┆ 0.914541        ┆ 0.000008      ┆ 0.914541                      ┆ NaN                    │
│ 2025-03-18 16:47:16.390 ┆ 100.0       ┆ 0.760916        ┆ 0.000011      ┆ 0.760916                      ┆ NaN                    │
│ …                       ┆ …           ┆ …               ┆ …             ┆ …                             ┆ …                      │
│ 2025-03-18 17:11:05.086 ┆ 480.0       ┆ 0.254668        ┆ 0.00003       ┆ 0.254668                      ┆ NaN                    │
│ 2025-03-18 17:12:17.779 ┆ 500.0       ┆ 0.256077        ┆ 0.00003       ┆ 0.256077                      ┆ NaN                    │
│ 2025-03-18 17:13:38.820 ┆ 520.0       ┆ 0.251833        ┆ 0.00003       ┆ 0.251833                      ┆ NaN                    │
│ 2025-03-18 17:14:48.713 ┆ 540.0       ┆ 0.246548        ┆ 0.00003       ┆ 0.246548                      ┆ NaN                    │
│ 2025-03-18 17:16:09.371 ┆ 560.0       ┆ 0.242846        ┆ 0.00003       ┆ 0.242846                      ┆ NaN                    │
└─────────────────────────┴─────────────┴─────────────────┴───────────────┴───────────────────────────────┴────────────────────────┘


metrics_system_df shape: (28, 3)
┌─────────────────────────┬──────────┬─────────────┐
│ weathergen.timestamp    ┆ perf.gpu ┆ perf.memory │
│ ---                     ┆ ---      ┆ ---         │
│ datetime[ms]            ┆ f64      ┆ f64         │
╞═════════════════════════╪══════════╪═════════════╡
│ 2025-03-18 16:42:37.571 ┆ 100.0    ┆ 64.0        │
│ 2025-03-18 16:43:35.623 ┆ 100.0    ┆ 60.0        │
│ 2025-03-18 16:44:47.891 ┆ 100.0    ┆ 60.0        │
│ 2025-03-18 16:46:03.834 ┆ 100.0    ┆ 66.0        │
│ 2025-03-18 16:47:16.390 ┆ 100.0    ┆ 66.0        │
│ …                       ┆ …        ┆ …           │
│ 2025-03-18 17:11:05.086 ┆ 100.0    ┆ 56.0        │
│ 2025-03-18 17:12:17.779 ┆ 100.0    ┆ 64.0        │
│ 2025-03-18 17:13:38.820 ┆ 100.0    ┆ 61.0        │
│ 2025-03-18 17:14:48.713 ┆ 100.0    ┆ 65.0        │
│ 2025-03-18 17:16:09.371 ┆ 100.0    ┆ 60.0        │
└─────────────────────────┴──────────┴─────────────┘
```

### CODE REVIEW

File: src/weathergen/utils/plot_training.py [hunk 1]
```diff
@@ -321,30 +334,23 @@ def plot_loss_per_run(
         clean_out_folder()
 
     runs_ids = {
-        "g0c86abp": [34298989, "ERA5 test"],
+        "thfmdl6w": [34298989, "ERA5 test"],
     }
 
     runs_data = [TrainLogger.read(run_id) for run_id in runs_ids]
 
+    # should be moved to the train logger
```
Comment (clessig) [inline src/weathergen/utils/plot_training.py]: @tjhunter  : this should go to train_logger.py

**Comment by clessig** [comment]
I enabled a minimal example. Question: can we have 'ERA5' with this captialization in the df cols?

More work to be needed but maybe you can already continue a few things from here.

File: src/weathergen/utils/plot_training.py [hunk 2]
```diff
@@ -162,31 +156,30 @@ def plot_loss_per_stream(
...
             for err in errs:
-                idx = 0 if mode == "train" else 1
                 linestyle = "-" if mode == "train" else ("--x" if len(modes) > 1 else "-x")
                 linestyle = ":" if "stddev" in err else linestyle
                 alpha = 1.0
                 if "train" in modes and "val" in modes:
                     alpha = 0.35 if "train" in mode else alpha
 
                 for j, (run_id, run_data) in enumerate(zip(runs_ids, runs_data, strict=False)):
-                    x_idx = [i for i, c in enumerate(run_data[idx][0]) if x_axis in c][0]
```
Comment (clessig) [inline src/weathergen/utils/plot_training.py:173]: @tjhunter : this selects the column to use for the x-axis (samples or reltime currently) by pattern matching the df col name with the specified pattern. Maybe there's a better way to do this using the df?

Comment (tjhunter) [reply src/weathergen/utils/plot_training.py:173]: I suggested a change

**Review by clessig — APPROVED** [review]
Looks good to me. The PR now also fixes some code breakages that were introduces in the last merge to develop so we should get this merged ASAP.
